dlme_airflow/drivers/json.py

```diff
@@ -0,0 +1,148 @@
+import pandas
+import logging
+import requests
+import jsonpath_ng
+
+from intake.source.base import DataSource, Schema
+
+
+class JsonSource(DataSource):
+    """
+    JsonSource lets you configure your Intake catalog entry with JSONPath
+    expressions that populate the resulting Pandas DataFrame.
+    See https://github.com/h2non/jsonpath-ng#jsonpath-syntax for what constitues a valid JSONPath.
+
+    For example if your Intake catalog contains:
+
+        args:
+            collection_url: https://www.loc.gov/collections/persian-language-rare-materials/?c=100&fo=json
+        metadata:
+          record_selector: "content.results"
+          fields:
+            id:
+              path: "id",
+            title:
+              path: "title"
+            thumbnail:
+              path: "resources[0].image"
+
+    and the collection URL returns JSON that looks like:
+
+    {
+      "content": {
+        "results": [
+          {
+            "id": "http://www.loc.gov/item/2017498321/",
+            "title": "[Majmuʻah, or, Collection]",
+            "resources": [
+              "image": "https://tile.loc.gov/image-services/iiif/service:amed:plmp:m154:0334/full/pct:6.25/0/default.jpg"
+            ]
+          },
+          {
+            "id": "http://www.loc.gov/item/00313408/",
+            "title": "Sakīnat al-fuz̤alāʼ mawsūm bi-ism-i tārīkhī-i Bahār-i Afghānī",
+            "resources": [
+              "image": "https://tile.loc.gov/image-services/iiif/service:amed:amedpllc:00406537456:0055/full/pct:6.25/0/default.jpg"
+            ]
+          }
+        ]
+      }
+    }
+
+    You would end up with a CSV with three columns (id, title and thumbnail) and two rows.
+    """
+
+    container = "dataframe"
+    name = "custom_json"
+    version = "0.0.1"
+    partition_access = True
+
+    def __init__(self, collection_url, metadata={}, csv_kwargs={}):
+        super(JsonSource, self).__init__(metadata=metadata)
+        self.collection_url = collection_url
+        self.metadata = metadata
+        self.csv_kwargs = csv_kwargs
+        self.record_limit = self.metadata.get("record_limit", None)
+        self.record_count = 0
+        self._setup_json_paths()
+
+    def read(self):
+        self._load_metadata()
+        df = self._get_dataframe()
+        if self.record_limit:
+            return df.head(self.record_limit)
+        else:
+            return df
+
+    def _get_schema(self):
+        return Schema(
+            datashape=None,
+            dtype=self.csv_kwargs.get("dtype"),
+            shape=None,
+            npartitions=1,
+        )
+
+    def _get_dataframe(self):
+        resp = requests.get(self.collection_url)
+        if resp.status_code != 200:
+            raise Exception(
+                f"HTTP request for {self.collection_url} resulted in {resp.status_code}"
+            )
+        return self._process_json(resp.json())
+
+    def _process_json(self, data):
+        result = self.record_selector.find(data)
```

Comment:
nitpicky, but what about calling this var `results` just so it's clearer that it's a list?

Reply:
Good point, especially because there's a result in this scope that reuses it.